devito/operations/interpolators.py

```diff
@@ -305,106 +305,64 @@ def callback():
-        ----------
-        expr : expr-like
-            Input expression to interpolate.
-        offset : int, optional
-            Additional offset from the boundary.
-        increment: bool, optional
-            If True, generate increments (Inc) rather than assignments (Eq).
-        """
-        def callback():
-            _expr = indexify(expr)
+    """
+    Concrete implementation of GenericInterpolator implementing a Linear interpolation
+    scheme, i.e. Bilinear for 2D and Trilinear for 3D problems.
 
-            p, _, _ = self.obj.interpolation_coeffs.indices
-            dim_subs = []
-            coeffs = []
-            for i, d in enumerate(self.obj.grid.dimensions):
-                rd = DefaultDimension(name="r%s" % d.name, default_value=self.r)
-                dim_subs.append((d, INT(rd + self.obj.gridpoints[p, i])))
-                coeffs.append(self.obj.interpolation_coeffs[p, i, rd])
-            # Apply optional time symbol substitutions to lhs of assignment
-            lhs = self.obj.subs(self_subs)
-            rhs = prod(coeffs) * _expr.subs(dim_subs)
+    Parameters
+    ----------
+    sfunction: The SparseFunction that this Interpolator operates on.
+    """
 
-            return [Inc(lhs, rhs)]
+    @cached_property
+    def _weights(self):
+        return {d: [1 - p/d.spacing, p/d.spacing]
+                for (d, p) in zip(self._gdim, self._psym)}
 
-        return Interpolation(expr, offset, increment, self_subs, self, callback)
+    def _coeff_temps(self, implicit_dims):
+        pmap = self.sfunction._position_map.values()
+        return [Eq(self._psym[d], pos - d.spacing*INT(floor(pos/d.spacing)),
+                   implicit_dims=implicit_dims)
+                for (d, pos) in zip(self._gdim, pmap)]
 
-    def inject(self, field, expr, offset=0):
-        """
-        Generate equations injecting an arbitrary expression into a field.
 
-        Parameters
-        ----------
-        field : Function
-            Input field into which the injection is performed.
-        expr : expr-like
-            Injected expression.
-        offset : int, optional
-            Additional offset from the boundary.
-        """
-        def callback():
-            _expr = indexify(expr)
-            _field = indexify(field)
-
-            p, _ = self.obj.gridpoints.indices
-            dim_subs = []
-            coeffs = []
-            for i, d in enumerate(self.obj.grid.dimensions):
-                rd = DefaultDimension(name="r%s" % d.name, default_value=self.r)
-                dim_subs.append((d, INT(rd + self.obj.gridpoints[p, i])))
-                coeffs.append(self.obj.interpolation_coeffs[p, i, rd])
-            rhs = prod(coeffs) * _expr
-            _field = _field.subs(dim_subs)
-            return [Inc(_field, rhs.subs(dim_subs))]
+class PrecomputedInterpolator(WeightedInterpolation):
 
-        return Injection(field, expr, offset, self, callback)
+    def _positions(self, implicit_dims):
```

Comment:
aren't `implicit_dims` an instance property? do we actually need to pass them or can we turn this into a property? 

btw, some of these properties *might* benefit from `cached_property`

Reply:
`implicit_dims` is an optional `kwarg` of `.interpolat/inject` in case the user wants to make it conditional so this has to process the user provided implicit dims in addition to the default sparse function dimensions. So cannot be made a property only